### ISSUE TITLE
22s cold start for hello world image

### ISSUE THREAD

**Issue by MaksymShuldiner** [microsoft/azure-container-apps#997]
> Please provide us with the following information:
> ---------------------------------------------------------------

### This issue is a: (mark with an x)
- [x] bug report -> please search issues before submitting

### Issue description
After creating Azure Container App from the default Hello World image, on the cold start, it takes 22 seconds to get a response. It's insane, because I had the same issue for my web app and thought that's because of container. 

![image](https://github.com/microsoft/azure-container-apps/assets/61598586/38ed600f-17a5-4848-9c75-67f024b66e11)

### Steps to reproduce
1. Create Azure Container App from default image
2. Track time for the cold start

**Expected behavior** - Cold start for the such small container should be 1-2 seconds

**Actual behavior** 22 seconds

**Comment by MaksymShuldiner** [comment]
also, I've tried launching 70MB node.js container and delay is still 20s. What can be a reason?

**Comment by MaksymShuldiner** [comment]
![image](https://github.com/microsoft/azure-container-apps/assets/61598586/7cd0f219-cb64-4990-baf5-e372a89490e6)
However, logs shows minimal delay in container pull and start

**Comment by guibranco** [comment]
Also having this issue with .NET 7 APIs

**Comment by MaksymShuldiner** [comment]
> Also having this issue with .NET 7 APIs

actually I've redeployed container into Cloud Run and got cold start as 2.5 seconds. So it's not definitely problem of container.

**Comment by howang-ms** [comment]
We are investigating this.

**Comment by edgarhuichen** [comment]
Same here. I have a simple python app and it takes around 30s to start the container and return a simple response.

**Comment by howang-ms** [comment]
For a quick update: We are working on the improvement, will give update as soon as the fix rollout.

**Comment by adeturner** [comment]
Looking forward to the fix. My 10 seconds example below with http scaler: 13:13:37 to 13:13:47

```txt
ContainerAppConsoleLogs_CL
| where RevisionName_s == "aca-weu-example--q5iwjne"

2023-12-20T13:13:47.8338291Z listening on port 80


ContainerAppSystemLogs_CL
| where RevisionName_s == "aca-weu-example--q5iwjne"

2023-12-20 13:19:07 +0000 UTC Deactivated apps/v1.Deployment k8se-apps/aca-weu-example--q5iwjne from 1 to 0
2023-12-20 13:19:07 +0000 UTC Stopping container examplecontainerapp
2023-12-20 13:19:07 +0000 UTC Pulling image "mcr.microsoft.com/azuredocs/containerapps-helloworld:latest"
2023-12-20 13:13:40 +0000 UTC Successfully pulled image "mcr.microsoft.com/azuredocs/containerapps-helloworld:latest" in 54.761604ms (54.768045ms including waiting)
2023-12-20 13:13:40 +0000 UTC Created container examplecontainerapp
2023-12-20 13:13:40 +0000 UTC Started container examplecontainerapp
2023-12-20 13:13:40 +0000 UTC Replica 'aca-weu-example--q5iwjne-7fb6fdd858-wdr69' has been scheduled to run on a node.
2023-12-20 13:13:37 +0000 UTC readiness probe failed: connection refused
```

**Comment by omni-htg** [comment]
Hello team! Has there been any good news regarding this fix?

Thanks!

**Comment by fengwusheng** [comment]
Google Cloud Run can set container to Gen1 to make cool start fast, only about 2 seconds, the container could be ready to use. We do not always have a newer version, Azure do not have to pull the image every time, make it an option such as "manually version pull" to make cool start easier and faster. Or option like Gen1 Gen2. The most important thing is: cool start faster, container should be ready in 2-3 seconds for small image.

**Comment by mluiten** [comment]
Hi @howang-ms -- any updates on the improvement 3 months later? Seeing the same. Small 50MB native docker image, pull time of 1-2 seconds, but total time for the first response is somewhere between 15-25 seconds. I love the premise of "scale to zero", but this makes it practically unusable, even when using small, fast images and runtimes.

**Comment by tbaroti** [comment]
Hi @howang-ms, @SophCarp,
are there any updates here? As @mluiten mentioned before, having the possibility to scale to zero (one of the mainly advertised feature of ACA) is not so useful if a cold start of 20-30 seconds is expected even for the smallest workloads. Even if we can't expect a quick solution, some update would be highly appreciated. Thank you!

**Comment by 03eltond** [comment]
This is a somewhat serious concern for us as well. For scale to zero, it can actually take well over a minute for us since our web server, once finally started up, then calls out to other backend containers that then have to spin up. Add to this the 2-3 minutes for node startup time on dedicated workloads and it's a pretty serious concern. Even for production loads, waiting 20-30 seconds for containers while scaling is proving to be very negative on our load testing compared to on prem results. We don't expect 1:1 for on prem since they are always on, but we expect to be able to meet demand in a reasonably timely manner when traffic is bursting.

**Comment by zyofeng** [comment]
The newly introduced workload profile does not address this unfortunately and is seriously restricting the adoption of container apps to basically async workloads (messaging, background etc).

**Comment by mluiten** [comment]
@howang-ms so, a few more months has passed without any additional information -- please let us know if you put this on the back-burner, or if no additional improvements are possible so we can start to look for other ways to improve the situation (for example; we're looking if we can call a fire-and-forget "ping" endpoint in the loading screen of our app). Either way, I would love some communication about this; the silence is deafening.

**Comment by timbodeit** [comment]
@torosent @microsoftopensource
Hong Wang doesn't seem to be responding to the question if there is any update available for several months now. Is there anyone else you could refer us to that would be able to provide an update on where this issue stands?

@howang-ms [mentioned](https://github.com/microsoft/azure-container-apps/issues/997#issuecomment-1843966596) you were working on an improvement. However I couldn't find a corresponding entry on the [roadmap](https://github.com/orgs/microsoft/projects/540/views/1).

Fully agree with @mluiten. Scale to zero is practically unusable for any services required for user-interactive integration at this time. Not knowing what sort of performance improvements we can expect and how long it will take for these to ship makes it hard to make an informed business decision whether Azure Container Apps are the right fit for our projects or if we should focus on alternatives like Google Cloud Run instead.

**Comment by cachai2** [comment]
Hi folks, thank you for your patience and for the continued feedback on this issue. We understand the importance of reducing cold start times for your applications and feel your pain. The changes that Hong had mentioned should have made ~7s improvement, but we are investigating this issue and the delta. We will provide an update once the investigation is complete. We do have more improvements in the pipeline for both cold start and the overall performance of the platform that we will be prioritizing. I do want to set expectations accordingly though and ask that you continue to be patient as the types of changes that we need to make for improving cold start are not small work items and will take time to implement and rollout. We can't currently commit to timelines, and major improvements will most likely not occur in the near term.
 
In the near term, for workloads that need to be much more responsive when scaling out today, we'd recommend you configure your minimum replica count for the application to be 1 or greater. This will ensure that your apps are always ready to go, and when inactive, these replicas will be billed at much lower idle rates per our [pricing page](https://azure.microsoft.com/en-us/pricing/details/container-apps/). Workarounds like what @mluiten mentioned about pinging the endpoint of the app will also work. However, due to the nature of billing for Azure Container Apps, the bill may be higher than setting the minimum replica count as pinging the application will make it become active vs inactive pricing if the app is set at a minimum and hasn't received traffic.
 
As folks have expressed on this thread, we can do a better job as a team of keeping you up to date on performance improvements that we are making. I've setup a recurring reminder on my calendar to update this thread when there are major improvements we have made. 

Finally, please continue to provide feedback here on your specific scenarios and the cold start times you are running into. The details provided from specific tests done by @MaksymShuldiner, @pollaktamas, @adeturner, and others on the thread are valuable for our understanding of the types of scenarios and workloads which are impacted. We are aiming to make general cold start improvements, but if there are specific changes we can make for certain high demand workloads, your feedback will be invaluable in prioritizing. It will also help to keep us honest, if the improvements you see in your production applications don't match the scale of improvements we claim, we want to know. If folks want to provide more direct feedback and discuss, I'd also be happy to chat about performance on Azure Container Apps. Feel free to reach out to me at cachai [at] microsoft [dot] com with details on your scenarios.

**Comment by cachai2** [comment]
Hi folks, providing a quick update as mentioned above. We've completed our investigations. The following numbers are for the quickstart image scenario without using managed identity. We've seen ~7s improvement at the P50 based on the changes Hong mentioned from ~22s to ~15s. The P95 after improvements was ~20s. In the last two weeks, we were able to also identify a quick fix that we have since implemented which has further improved these numbers to ~10s at the P50- and ~15s at the P95. Feel free to validate these numbers on your own and let us know if you are seeing any differences. Please note that these numbers are based on the quickstart image so depending on your images, dependencies, etc, you will most likely see higher cold start numbers for your actual app workloads. 

We do have further improvements we are working on. However, as mentioned in my previous message, these changes will be further out, and we don't currently have an ETA on when they will land. In the meantime, please continue to provide feedback on this area, and let us know what cold start numbers you are seeing for your scenarios. Thank you!

**Comment by GibreelAbdullah** [comment]
The cold start time today is about 12s for a node app. The same image is deployed in cloud run where the cold start time remains at around 2-3s.
Reading all the comments, there has been a significant progress (from 25s to 12s) but much more improvement is still needed.

**Comment by ctigrisht** [comment]
I wanted to stay on azure but this makes it impossible, GCP run gets a response within 2 seconds on a cold start, the last I checked on azure it was 14-18 seconds for a nodejs image and a bit more for an ASP.NET image

**Comment by andrew-vdb** [comment]
Azure is never good with cold start.
There is business issue than technical as they want us to keep paying with at least 1 instance

**Comment by deanhdev** [comment]
Hi, 

this slow cold start is also a problem with azure container app **jobs**.
Container App Jobs is a perfect fit for our needs (spinning up longer running background tasks on demand), but it also suffers from long(er) cold starts. It is also different per job. Sometimes it takes more or less long, also the container image pull takes different times per job (maybe this could be fixed with a higher tier of container registry? I did not find any information regarding network speed by tier).

Are there any updates on this?

**Comment by caledhwa-pnnl** [comment]
To pile on: I'm running an Azure Container App Job every 5 minutes which is a "hello world" which is a dotnet 8 simple exe which outputs like 6 console lines and that's all. Each execution is basically 30-40 seconds. Seems like a lot of spin up time for an exe.

**Comment by aleroca96** [comment]
hi - wanted to check if  there are any updates on this topic?

We are loving ACA on our firm and currently have a rule to have a minimum replica of 1 between working hours, but it'd be ideal if cold start times could be reduced.

Thanks!

**Comment by 0xkalle** [comment]
Any news here?

**Comment by tiaringhio** [comment]
bumping, we've implemented the same solution as @aleroca96 because scaling to 0 during working hours is unfeasible as of right now.

**Comment by iSaulX** [comment]
Two years later and we haven't received any fix for this. Cold stars still take 20 seconds or even more. What's the point of serverless if you can't scale down to zero?

**Comment by 0xkalle** [comment]
Exactly. I tried optimizing the docker images, made sure to use the right registry etc..but it all has minimal effect compared to the base start time...

So only thing left Todo is not scale to zero until I migrate away.

**Comment by TheCodingOwl** [comment]
We're having the same issue. According to the logs, everything should be relatively fine within just a few seconds. But the first request I make still takes 20 seconds. I'm not sure how to diagnose what's even happening behind the scenes.

**Comment by boldandbusted** [comment]
> We're having the same issue. According to the logs, everything should be relatively fine within just a few seconds. But the first request I make still takes 20 seconds. I'm not sure how to diagnose what's even happening behind the scenes.

Yeah, that should be part of the benefit of using Azure Container Apps (or any api-driven Cloud Native solution): we don't need to care what's happening behind the scenes/under the hood. But, that implies that the service is high quality. :(

Some communication would be most welcome from Microsoft here. We should see at least some discussion on the sources of the delays, so we can tag along for the troubleshooting ride. ;)

**Comment by espenairmine** [comment]
Any update on this?

**Comment by rexhinokovaci** [comment]
Really Annoying for ACA, whole idea for ACA is to be scaling to 0 and it takes more than 30s to scale to 1.

**Comment by Ily83** [comment]
it's not only with hello world image

**Comment by ricoprzybylski** [comment]
+1 for this. Really annoying issue. Whole beauty of serverless programming is broke here. I am not able to setup cheap deployment. If on production my login request take up to 20-25s form cold start I am force to leave every service at least 1 instance. Services are running generating costs but 95% are just waiting ...

**Comment by FlorentATo** [comment]
Hello @cachai2, I have 3 quick questions:
1. Is there any update on this issue ? Looking at the roadmap, I see it still "In-progress" so I assume the PG team is still working on it ?
2. Cloud you clarify when exactly does the billing start ? I want to be sure your customers aren't paying for resource utilization that they aren't actually using.

Please keep us posted 🙏 
cc @howang-ms, @SophCarp

**Comment by mohamedmaged** [comment]
Strongly agree this needs to be prioritized. Cold start is one issue, but the bigger problem for us is scale-up latency under loadby the time new replicas are ready, the traffic spike is already over and existing replicas have been overwhelmed.

At this speed it's not really possible to match live production traffic patterns without permanently over-provisioning, which defeats much of the value proposition of a serverless container platform. Would love an update on where this sits on the roadmap.

**Comment by GABRIELNGBTUC** [comment]
What is strange is that azure flex consumption uses the same backend as container apps (Project legion) and everything works as expected there.

HTTP call cold start on a function app running on flex consumption vs container apps takes <5s vs 40+s for the same container image and scale out is near instant on flex.

Perhaps the ACA team can look at the flex consumption implementation to see how they can improve things because today, Azure function on container apps with scale to 0 is borderline useless compared to flex.

**Comment by boldandbusted** [comment]
> What is strange is that azure flex consumption uses the same backend as container apps (Project legion) and everything works as expected there.
> 
> HTTP call cold start on a function app running on flex consumption vs container apps takes <5s vs 40+s for the same container image and scale out is near instant on flex.
> 
> Perhaps the ACA team can look at the flex consumption implementation to see how they can improve things because today, Azure function on container apps with scale to 0 is borderline useless compared to flex.

That is a very interesting observation.

It is only a guess, but I suppose that in Flex they may "cheat" a bit and keep a minimum, finiite number of instances of an application running or in suspension in the background in the absence of active requests. In other words, a "primed" backend pool for a "warm start". 

Cheers.

**Comment by FlorentATo** [comment]
> What is strange is that azure flex consumption uses the same backend as container apps (Project legion) and everything works as expected there.
> 
> HTTP call cold start on a function app running on flex consumption vs container apps takes <5s vs 40+s for the same container image and scale out is near instant on flex.
> 
> Perhaps the ACA team can look at the flex consumption implementation to see how they can improve things because today, Azure function on container apps with scale to 0 is borderline useless compared to flex.

Could we have someone from the Container apps PG team comment on this analysis ? 👀

**Comment by tiaringhio** [comment]
guys, this announcement was released 2 days ago

[Introducing Container App Express](https://techcommunity.microsoft.com/blog/appsonazureblog/introducing-azure-container-apps-express/4519150)

some of the highlights:

<img width="761" height="135" alt="Image" src="https://github.com/user-attachments/assets/a9265175-1336-414f-b215-a3a343ca97a8" />

does that mean we are finally going to see some improvements in cold starts?

**Comment by yoDon** [comment]
It's great to see someone working on Express Container App Environments as a way to improve the cold start time, but please please prioritize Managed Identity support for Express Container App Environments. 

We do not want to be baking our ACR credentials into our deployments, nor do we want to be making our containers public. The rest of Azure was built around managed identity support. Express needs to support Managed Identity too.

As claude puts it, `When your container app starts (or scales up), the Azure Container Apps platform needs to pull your image from ACR. In full ACA: The container app has a managed identity and uses that to authenticate to ACR. In Express: No managed identity means the container app needs stored credentials in its configuration.`

**Comment by ajmawer** [comment]
After today's ACR outage, I hope somewhere in the mix for this is a caching mechanism so that a service doesn't completely lose its ability to scale when ACR is offline...
